ca/ca.go

```diff
@@ -218,3 +219,44 @@ func New(log *log.Logger, db *db.MemoryStore) *CAImpl {
 	}
 	return ca
 }
+
+func (ca *CAImpl) CompleteOrder(order *core.Order) {
+	// Check the authorizations - this is done by the VA before calling
+	// CompleteOrder but we do it again for robustness sake.
+	for _, authz := range order.AuthorizationObjects {
+		// Lock the authorization for reading
+		authz.RLock()
+		if authz.Status != acme.StatusValid {
+			return
+		}
+		authz.RUnlock()
+	}
+
+	order.RLock()
```

Comment:
You need to lock earlier, before you fetch order.AuthorixationObjects, Also I think a full lock would be fine; we're not worried about performance here, and there's no deadlock possibility, right?